### PR TITLE
Scripts/Spells: Mine Sweeper achievement

### LINKED ISSUE #15938: Achievement Mine Sweeper
Achievement [Mine Sweeper](http://wotlk.openwow.com/achievement=1428) doesn't work.
"Get caught in 10 consecutive land mine explosions in the Sparksocket Minefield without landing."
[Criteria spell](http://wotlk.openwow.com/spell=57064)

hash f8f815dbc3a9
branch 3.3.5

### PR DESCRIPTION
**Pull request by ghost**
Enable achievement 1428, "Mine Sweeper" in SpellScript

by @Re3os / @ccrs / @sirikfoll . based on original script from @Vincent-Michael

Closes #15938

---

Updated 2015-12-15 04:27:17 +0100

### CODE REVIEW

**Comment by Shauren** [comment]
Absolutely not. CompletedAchievement method must never be used, it exists only for use in GM commands.

**Comment by tkrokli** [comment]
OK, I did not know that. I suppose ccrs and Killyana did not know either. But have you got any suggestion for what could be missing from either Vincent-Michael's original script or the revamped one I tested in  #15938 that did not work, even with the suggested improvements from ccrs? ... or any other suggestion for what can be done to make such a script work?

**Comment by MitchesD** [comment]
This achievement is easy to fix, I have complete fix somewhere saved, but there is one core issue which must be solved before it can be scripted properly. And the problem is: when you are on the mine, gobject should summon (or it maybe already does) some trigger which should deal you damage and some aura, however currently that trigger (dummy) is friendy, because it's player invoked action I guess, and that's the problem, that trigger should be unfriendly to be able to give you that aura, damage and knockback.

**Comment by tkrokli** [comment]
OK, I thought that was taken care of by the SQL part:
```sql
UPDATE `creature_template` SET `flags_extra`= 130 WHERE `entry`= 29397;
```
http://www.wowhead.com/npc=29397/land-mine-bunny#comments
This bunny is on every landmine in the field. With GM ON it is visible, but only at close range (3-5 m).

But maybe the creature entry or the flag is wrong, I forgot to check that bit in unchanged source.
Although for every landmine explosion, the character entered combat and left combat right after.
Will test again with the previous version & straight source + check the bunnies' creature entry & flags.
If something useful can be changed there, maybe it will be possible to make it work with this
```cpp
target->CastSpell(target, SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT, true);
```
instead of the current invalid code, after all. I need to check that closer.

**Comment by MitchesD** [comment]
Read my comment again to prevent these long comments. It's a core bug. It doesn't help if you change any flag

**Comment by tkrokli** [comment]
OK. So, close this PR, then?

File: src/server/scripts/Spells/spell_generic.cpp [hunk 1]
```diff
@@ -4120,6 +4120,45 @@ class spell_gen_mixology_bonus : public SpellScriptLoader
...
+            void HandleScript(SpellEffIndex /*effIndex*/)
+            {
+                if (Player* target = GetHitPlayer())
+                {
+                    Aura const* aura = GetHitAura();
+                    if (!aura || aura->GetStackAmount() != 10)
+                        return;
+
+                    AchievementEntry const* achievement = sAchievementStore.LookupEntry(LANDMINE_KNOCKBACK_ACHIEVEMENT);
+                    target->CompletedAchievement(achievement);
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_generic.cpp]: hack

File: src/server/scripts/Spells/spell_generic.cpp [hunk 2]
```diff
@@ -4120,6 +4120,45 @@ class spell_gen_mixology_bonus : public SpellScriptLoader
...
+                        return;
+
+                    target->CastSpell(target, SPELL_LANDMINE_KNOCKBACK_ACHIEVEMENT, true);
+                }
+            }
+
+            void Register() override
+            {
+                OnEffectHitTarget += SpellEffectFn(spell_gen_landmine_knockback_achievement_SpellScript::HandleScript, EFFECT_0, SPELL_EFFECT_SCRIPT_EFFECT);
+                OnEffectHitTarget += SpellEffectFn(spell_gen_landmine_knockback_achievement_SpellScript::HandleScript, EFFECT_1, SPELL_EFFECT_APPLY_AURA);
```
Comment (sirikfoll) [inline src/server/scripts/Spells/spell_generic.cpp]: this line is not necessary:
`OnEffectHitTarget += SpellEffectFn(spell_gen_landmine_knockback_achievement_SpellScript::HandleScript, EFFECT_1, SPELL_EFFECT_APPLY_AURA);`

File: src/server/scripts/Spells/spell_generic.cpp [hunk 3]
```diff
@@ -4120,6 +4120,45 @@ class spell_gen_mixology_bonus : public SpellScriptLoader
...
+            }
+        };
+
+    SpellScript* GetSpellScript() const override
```
Comment (sirikfoll) [inline src/server/scripts/Spells/spell_generic.cpp]: indentation?

File: sql/updates/world/2015_12_10_14_world.sql [hunk 4]
```diff
@@ -0,0 +1,9 @@
+--
+-- Landmine Knockback Achievement (29397,54355,54402,57099)
+DELETE FROM `spell_linked_spell` WHERE `spell_trigger` = 54355;
+INSERT INTO `spell_linked_spell` (`spell_trigger`, `spell_effect`, `comment`) VALUES
+(54355, 54402, 'Landmine Knockback Achievement');
+
```
Comment (sirikfoll) [inline sql/updates/world/2015_12_10_14_world.sql]: the above lines are not necessary ( already are in the current TDB)

Comment (tkrokli) [inline sql/updates/world/2015_12_10_14_world.sql]: OK, but isn't entry 5258 a Cataclysm achievement? (I will change this anyway, just curious.)

File: src/server/scripts/Spells/spell_generic.cpp (hunk 3 above)
Comment (tkrokli) [inline src/server/scripts/Spells/spell_generic.cpp]: Could you be a bit more specific? :) Should there be less? And which lines? I used the 4 space standard.

File: src/server/scripts/Spells/spell_generic.cpp (hunk 2 above)
Comment (tkrokli) [inline src/server/scripts/Spells/spell_generic.cpp]: OK, will remove it. It was more or less a test if it would help in troubleshooting.

**Comment by sirikfoll** [comment]
After updated, I think this can be merged. It's working perfectly.
the core issue mentioned by @MitchesD isn't affecting this because there is already a hack in tdb avoiding that.

when the core issue is resolved, simply add this:
```sql
DELETE FROM `spell_linked_spell` WHERE `spell_trigger`=54355;
UPDATE `creature_template` SET `AIName`='SmartAI' WHERE `entry`=29397;
DELETE FROM `smart_scripts` WHERE `entryorguid`=29397 AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(29397,0,0,0,54,0,100,0,0,0,0,0,11,54402,0,0,0,0,0,1,0,0,0,0,0,0,0,'Land Mine Bunny - On Reset - Cast \'Land Mine Knockback\'');
```
or not, will work either way ^^

File: sql/updates/world/2015_12_10_14_world.sql (hunk 4 above)
Comment (sirikfoll) [inline sql/updates/world/2015_12_10_14_world.sql]: 5258 is the criteria_id for the achievement 1428 ^^

File: src/server/scripts/Spells/spell_generic.cpp (hunk 3 above)
Comment (sirikfoll) [inline src/server/scripts/Spells/spell_generic.cpp]: 4 more spaces in this line and in the next 3 lines

File: src/server/scripts/Spells/spell_generic.cpp (hunk 2 above)
Comment (sirikfoll) [inline src/server/scripts/Spells/spell_generic.cpp]: will work

**Comment by tkrokli** [comment]
I don't know if this is 100% the same on the 6.x branch, because on wowhead.com the Land Mine Bunny, entry 29397(infernal model), is described as "**This NPC is not in game.**" in red letters.

**Comment by sirikfoll** [comment]
same thing, i've just sniffed this in retail ^^
also, this sql part is still needed
```sql
DELETE FROM `spell_script_names` WHERE `ScriptName`='spell_gen_landmine_knockback_achievement';
INSERT INTO `spell_script_names` (`spell_id`,`ScriptName`) VALUES
(57099,'spell_gen_landmine_knockback_achievement');

**Comment by tkrokli** [comment]
Right. I should have read your SQL comment better. On it now.

**Comment by MitchesD** [comment]
I wonder, how could it be tested @tkrokli if you didn't even added spellscript to your DB, test your fixes!

**Comment by sirikfoll** [comment]
that spell_linked is already in tdb 335.60

**Comment by tkrokli** [comment]
TrinityCore rev. 2937af2a5990 2015-12-13 01:14:08 +0100 (landmine_knockback_achievement branch) (Win64, Release) - confirmed. It actually works. I would not have thought that by just looking at the code alone. But it really works. - and yes, like @sirikfoll said, it already exists in TDB 335.60:
```sql
SELECT * FROM `spell_linked_spell` WHERE `spell_trigger` = 54355;
```
```
spell_trigger  spell_effect    type  comment
-------------  ------------  ------  -----------
        54355         54402       0  Land Mine
```

**Comment by sirikfoll** [comment]
For the dbError
```sql
DELETE FROM `achievement_criteria_data` WHERE `criteria_id`=5258;
INSERT INTO `achievement_criteria_data` (`criteria_id`, `type`, `value1`, `value2`, `ScriptName`) VALUES
(5258,0,0,0,'');

**Comment by tkrokli** [comment]
Thanks a lot. :) I was about to experiment with the values. At first, I thought `type = 29`, but I was wrong.

**Comment by tkrokli** [comment]
@MitchesD
I don't know why I did not check it before, but that `spell_linked_spell` `(54355, 54402)` part was added 2015-04-18, right after TDB 335.58 and before TDB 335.59, in commit https://github.com/TrinityCore/TrinityCore/commit/2023dd4 from https://github.com/TrinityCore/TrinityCore/issues/965#issuecomment-93392884 .

**Comment by tkrokli** [comment]
I would like to have comments on which parts makes this PR still [Feedback-FixOutdatedMissingWIP]
and not labeled [Feedback-PatchFix], so I can to improve it to be good enough for TC standards.

**Comment by ccrs** [comment]
everything seems correct, just normal random TC merge time xd

**Comment by tkrokli** [comment]
Yes, I am very well used to the merge time. My suggestion was only about setting the right label on it.

**Comment by jackpoz** [comment]
@tkrokli how much time did it take to earn the achievement ingame ?

**Comment by tkrokli** [comment]
With the max time of 2.5 seconds between the explosions, it should not take more than 30 seconds in total. Then again, if not using the level 77 "loaner" from K3 (or regular mount from level 80), it could take some time to get used to triggering the explosions correctly by running and then steering where the explosion throws the player character. The techniques are discussed in the wowhead comments.

**Pull request merged**